api/api-public-tenant-permissions.go

```diff
@@ -130,37 +103,164 @@ func (s *server) AddPermission(ctx context.Context, in *pb.AddPermissionRequest)
 	}
-	for _, permAction := range permissionObj.Actions {
-		pbAction := pb.PermissionAction{
-			Id:   permAction.ID.String(),
-			Type: string(permAction.ActionType),
+
+	// start updating values on the permission obj not yet on db.
+	permission.Name = permissionName
+	permission.Effect = effect
+	permission.Description = &description
+
+	// Nullified values if they are empty
+	if description == "" {
+		permission.Description = nil
+	}
+
+	// -- Update Permission RESOURCES
+	var currentResourceBucketNames []string
+	mapResourceName := make(map[string]string)
+	for _, perm := range permission.Resources {
+		currentResourceBucketNames = append(currentResourceBucketNames, perm.BucketName)
+		mapResourceName[perm.BucketName] = perm.ID.String()
+	}
+	// TODO: parallelize
+	resourcesToCreate := differenceArrays(resourcesBucketNames, currentResourceBucketNames)
+	resourcesToDelete := differenceArrays(currentResourceBucketNames, resourcesBucketNames)
+	// CREATE New Resources
+	// create a Temporal permission to Create the new Permission resources
+	tempPermission := &cluster.Permission{ID: permission.ID}
+	cluster.AppendPermissionResourcesObj(tempPermission, resourcesToCreate)
+	// for each resource, save to DB
+	for _, resc := range tempPermission.Resources {
+		err = cluster.InsertResource(appCtx, tempPermission, &resc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// DELETE unwanted Resources
+	for _, bucketName := range resourcesToDelete {
+		resourceID, ok := mapResourceName[bucketName]
+		if !ok {
+			return nil, status.New(codes.Internal, "error retrieving resource to delete").Err()
+		}
+		err = cluster.DeletePermissionResourceDB(appCtx, resourceID)
+		if err != nil {
+			return nil, status.New(codes.Internal, "error deleting existing resource").Err()
 		}
-		res.Actions = append(res.Actions, &pbAction)
 	}
 
-	if permissionObj.Description != nil {
-		res.Description = *permissionObj.Description
+	// -- Update Permission ACTIONS
+	var currentActionTypes []string
+	mapActionTypes := make(map[string]string)
+	for _, perm := range permission.Actions {
+		currentActionTypes = append(currentActionTypes, string(perm.ActionType))
+		mapActionTypes[string(perm.ActionType)] = perm.ID.String()
 	}
-	return res
+	// TODO: parallelize
+	actionsToCreate := differenceArrays(actionTypes, currentActionTypes)
+	actionsToDelete := differenceArrays(currentActionTypes, actionTypes)
+	// CREATE New Actions
+	cluster.AppendPermissionActionObj(tempPermission, actionsToCreate)
+	// for each resource, save to DB
+	for _, action := range tempPermission.Actions {
+		err = cluster.InsertAction(appCtx, tempPermission, &action)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// DELETE unwanted Actions
+	for _, types := range actionsToDelete {
+		actionID, ok := mapActionTypes[types]
+		if !ok {
+			return nil, status.New(codes.Internal, "error retrieving action to delete").Err()
+		}
+		err = cluster.DeletePermissionActionDB(appCtx, actionID)
```

Comment:
Do we support bulk delete instead of deleting the records one by one?

Reply:
done